### ISSUE TITLE
Pre-training info

### ISSUE THREAD

**Issue by nargenziano** [NVlabs/SegFormer#116]
Hello and thanks for the work.

I was wondering if you could share more info regarding the pre-training of MiT architectures. 
I've read from other issues that the configs are the same as pvt_v2, but what is the actual pre-training code you used? Is it the PVT classification training?
I tried to edit the PyramidVisionTransformer model to make it identical to MiT B3 and ran the ImageNet classification training of PVT from scratch, however, the classification performance was worse than the one expected for PVT-v2 B3 (around 77.3% Acc@1, instead of the expected 83.1%). What is the expected pre-training performance of MiT?

**Comment by gauenk** [comment]
The paper states "We pre-train the encoder on the Imagenet-1K dataset".

Does this mean the encoder is trained a classification task first? If so, is there code for this to share? I can not find it in the repo.

Primarily, I want to be able to reproduce the "mit_*pth" files, either conceptually or with your code.

**Comment by gauenk** [comment]
following up...

**Comment by wangh09** [comment]
Same question here. Seems that the classification head(commented out) in MiT backbone won't work cause the output of stage 4 is B*49*512, and can't directly be followed with an nn.Linear to output B*1000.

**Comment by Mike-HH** [comment]
following up..

**Comment by waw123456** [comment]
following up...

**Comment by davidhuangal** [comment]
Hey all, I know this is a bit of an old issue, but I was wondering the same thing myself.

In a couple of older issues, the first author of the paper says that MiT was trained in the same way as PvT v2:
* https://github.com/NVlabs/SegFormer/issues/97#issuecomment-1218640154
* https://github.com/NVlabs/SegFormer/issues/67#issuecomment-1024199226

And the code does look very similar, and I think I've found the solution.

* In PvT v2, the last output tensor, `x`, is of shape `(batch size x num tokens x embedding dimension)`

* Then an average across all tokens is calculated via `x.mean(dim=1)` (https://github.com/whai362/PVT/blob/57e2dfaa5a46f9050d76f306a4fcd9a7c061f520/classification/pvt_v2.py#L295)

* This results in a tensor of shape `(batch size x embedding dimension)` which can now be passed through a linear layer for classification.

And actually as I was searching the issues, I found the first author explicitly saying that this is what they did: https://github.com/NVlabs/SegFormer/issues/39#issuecomment-889703999

As for the training settings, they are listed in the [PvT paper](https://arxiv.org/abs/2102.12122):
> We follow DeiT [63] and apply random cropping, random horizontal flipping [59], label-smoothing regularization [60], mixup [78], CutMix [76], and random erasing [82] as data augmentations. During training, we employ AdamW [46] with a momentum of 0.9, a mini-batch size of 128, and a weight decay of 5 × 10−2 to optimize models. The initial learning rate is set to 1 × 10−3 and decreases following the cosine schedule [45]. All models are trained for 300 epochs from scratch on 8 V100 GPUs. To benchmark, we apply a center crop on the validation set, where a 224× 224 patch is cropped to evaluate the classification accuracy.


I think some of the confusion came from the fact that the `mix_transformer.py` code has the old classification head code commented out:
```python
 # classification head
 # self.head = nn.Linear(embed_dims[3], num_classes) if num_classes > 0 else nn.Identity()
```
but the averaging across tokens code was omitted.


In any case, unfortunately this does not explain the drop in performance that @nargenziano saw. Has anyone else tried reproducing MiT's ImageNet results?
